generic/sharded/sharded.go

```diff
@@ -0,0 +1,186 @@
+package sharded
+
+import (
+	"errors"
+	"time"
+
+	lazylru "github.com/TriggerMail/lazylru/generic"
+)
+
+// LazyLRU is a sharded version of the lazylru.LazyLRU cache. The goal of this
+// cache is to reduce lock contention via sharding. This may have a negative
+// impact on memory locality, so your mileage may vary.
+//
+// LazyLRU is an LRU cache that only reshuffles values if it is somewhat full.
+// This is a cache implementation that uses a hash table for lookups and a
+// priority queue to approximate LRU. Approximate because the usage is not
+// updated on every get. Rather, items close to the head of the queue, those
+// most likely to be read again and least likely to age out, are not updated.
+// This assumption does not hold under every condition -- if the cache is
+// undersized and churning a lot, this implementation will perform worse than an
+// LRU that updates on every read.
+type LazyLRU[K comparable, V any] struct {
+	sharder func(K) uint64
+	shards  []*lazylru.LazyLRU[K, V]
+	ttl     time.Duration
+}
+
+// New creates a new sharded cache with strings for keys and any (interface{})
+// for values
+//
+// Deprecated: To avoid the casting, use the generic NewT interface instead
+func New(maxItemsPerShard int, ttl time.Duration, numShards int) *LazyLRU[string, any] {
+	return NewT[string, any](maxItemsPerShard, ttl, numShards, StringSharder)
+}
+
+// NewT creates a new sharded cache. The sharder function must be consistent
+// and should be as uniformly-distributed over the expected source keys as
+// possible. For string and []byte keys, the pre-canned StringSharder and
+// BytesSharder are appropriate. These are both based on the HashingSharder,
+// which callers can use to create sharder functions for custom types.
+func NewT[K comparable, V any](maxItemsPerShard int, ttl time.Duration, numShards int, sharder func(K) uint64) *LazyLRU[K, V] {
+	shards := make([]*lazylru.LazyLRU[K, V], numShards)
+	for i := 0; i < numShards; i++ {
+		shards[i] = lazylru.NewT[K, V](maxItemsPerShard, ttl)
+	}
+
+	return &LazyLRU[K, V]{sharder, shards, ttl}
+}
+
+// ShardIx determines the target shard for the provided key
+func (slru *LazyLRU[K, V]) ShardIx(key K) int {
+	return int(slru.sharder(key)&0x7FFFFFFFFFFFFFFF) % len(slru.shards)
+}
+
+// IsRunning indicates whether the background reaper is active on at least one
+// of the shards
+func (slru *LazyLRU[K, V]) IsRunning() bool {
+	for _, s := range slru.shards {
+		if s.IsRunning() {
+			return true
+		}
+	}
+	return false
```

Comment:
The individual `IsRunning` calls are locked internally. Since we don't need an exact count, I don't think we need a global lock. Honestly, I'm not sure anyone will ever use this function in the real world. You'd start up the cache(s), do your thing, then shut down the process -- graceful shutdown won't really be a thing. If that seemed like a more likely use case, I'd probably make the `Close` call block.